datalad/cmd.py

```diff
@@ -98,7 +98,208 @@ def _cleanup_output(stream, std):
+                    # this will always be the full report so far,
+                    # not just an output increment
+                    pout = (poll.stdout, poll.stderr)
+
+                if proc_out is not None:
+                    for i, (o, proc) in enumerate(zip(pout, proc_out)):
+                        if proc is None:
+                            if o is not None:
+                                out[i] += o
+                            # no index update needed, can never change
+                            continue
+                        # current length of output
+                        len_o = len(o) if o else 0
+                        # anything new in the output?
+                        if len_o - last_idx[i] > 0:
+                            # engage output processed
+                            processed, unprocessed_len = proc(o[last_idx[i]:])
+                            # only keep the processed output
+                            out[i] += processed
+                            # record point to which we processed the output
+                            # subtract number of bytes reported as unprocessed
+                            last_idx[i] = len_o - unprocessed_len
+
+            # obtain exit code
+            status = process.poll()
+
+            # decode bytes to string
+            out = tuple(o.decode('utf-8') if o else '' for o in out)
+
+            if status not in [0, None]:
+                msg = "Failed to run %r%s." % (
+                    cmd,
+                    (" under %r" % self.cwd) if self.cwd else '',
+                )
+                raise CommandError(
+                    cmd=str(cmd),
+                    msg=msg,
+                    code=status,
+                    stdout=out[0],
+                    stderr=out[1],
+                )
+            else:
+                lgr.log(8, "Finished running %r with status %s", cmd, status)
+
+        except CommandError:
+            # do not bother with reacting to "regular" CommandError
+            # exceptions.  Somehow if we also terminate here for them
+            # some processes elsewhere might stall:
+            # see https://github.com/datalad/datalad/pull/3794
+            raise
+
+        except BaseException as exc:
+            exc_info = sys.exc_info()
+            # KeyboardInterrupt is subclass of BaseException
+            lgr.debug("Terminating process for %s upon exception: %s",
+                      cmd, exc_str(exc))
+            try:
+                # there are still possible (although unlikely) cases when
+                # we fail to interrupt but we
+                # should not crash if we fail to terminate the process
+                process.terminate()
+            except BaseException as exc2:
+                lgr.warning("Failed to terminate process for %s: %s",
+                            cmd, exc_str(exc2))
+            raise exc_info[1]
+
+        return out
+
+
+class Runner(WitlessRunner):
```

Comment:
Why make Runner() inherit from WitlessRunner()?

Reply:
Because it also handles `self.cwd` and `self.env` for both. If there is a reason to duplicate that, and could be convinced to revert this.

Reply:
IMO that's not a reason to inherit from a base class when there are large api conflicts between the parent and child.

Reply:
OK, will copy&paste the code.

Reply:
Reverted all changes to `Runner`